connectors/sources/jira.py

```diff
@@ -0,0 +1,467 @@
+            attachments (list): List of attachments for an issue
+            issue_key (str): Issue key for generating `_id` field
+
+        Yields:
+            Dictionary: Jira attachment on the given issue
+            attachment (dict): Attachment dictionary for extracting the content
+        """
+        for attachment in attachments:
+            yield {
+                "_id": f"{issue_key}-{attachment['id']}",
+                "title": attachment["filename"],
+                "type": "Attachment",
+                "issue": issue_key,
+                "_timestamp": attachment["created"],
+                "size": attachment["size"],
+            }, attachment
+
+    async def _grab_content(self, attachments, issue_key):
+        """Coroutine to add attachments to Queue and get content
+
+        Args:
+            attachments (list): List of attachments for an issue
+            issue_key (str): Issue key for generating `_id` field
+        """
+        async for content, attachment in self._get_attachments(
+            attachments=attachments, issue_key=issue_key
+        ):
+            if attachment:
+                await self.queue.put(
+                    (
+                        content,
+                        partial(
+                            self.get_content,
+                            issue_key=issue_key,
+                            attachment=copy(attachment),
+                        ),
+                    )
+                )
+
+    async def get_docs(self, filtering=None):
+        """Executes the logic to fetch jira objects in async manner
+
+        Args:
+            filtering (Filtering): Object of class Filtering
+
+        Yields:
+            dictionary: dictionary containing meta-data of the files.
+        """
+
+        async def _project_task():
+            """Coroutine to add projects documents to Queue"""
+            async for project_data in self._get_projects():
+                await self.queue.put((project_data, None))
+            await self.queue.put("FINISHED")
+
+        async def _document_task():
+            """Coroutine to add issues/attachments to Queue"""
+            async for document, issue in self._get_issues():
+                await self.queue.put((document, None))
+                attachments = issue["fields"]["attachment"]
+                if len(attachments) > 0:
+                    await self.fetchers.put(
+                        partial(self._grab_content, attachments, issue["key"])
+                    )
+            await self.queue.put("FINISHED")
+
+        project_task = asyncio.create_task(_project_task())
+        attachment_task = asyncio.create_task(_document_task())
+        completed_tasks = 0
+        while True:
+            _, item = await self.queue.get()
+            if item == "FINISHED":
+                completed_tasks += 1
+                if completed_tasks == 2:
+                    break
+            else:
+                yield item
```

Comment:
I'm confused with this function, why is it done this way? What is it trying to do?

Reply:
This is a part of Tarek's suggestion - https://github.com/elastic/connectors-python/issues/256#issuecomment-1367643403. This code is working as a consumer i.e. yielding documents to BYOC. In `get_docs` method, we've made two tasks for fetching projects and issues(including the attachments) and once both tasks are completed, consumer stops.

Reply:
I see. One small thing -  I think it's nice to test this logic if possible and document what this loop means - it takes quite a while to understand what it does tbh.

Reply:
Sure, I'll document it by adding a comment above this code hoping that's fine for you and add a test case for this logic.

Reply:
This logic is already covered in the `test_get_docs` test case.